rom/laghos_rom.cpp

```diff
@@ -338,6 +339,84 @@ void ComputeElementRowOfG_E(const IntegrationRule *ir,
     }
 }
 
+// Sets the rows of constraints matrix G for the energy-conserving EQP rule. 
+void ComputeRowsOfG(const IntegrationRule *ir,
+					hydrodynamics::QuadratureData const& quad_data,
+					Vector const& w_e, Vector const& v_e,
+					FiniteElement const& test_fe,
+					FiniteElement const& trial_fe,
+					const int zone_id,
+					const bool equationV, const bool equationE,
+					Vector & rv, Vector & re)
+{
+	const int nqp = ir->GetNPoints();
+	const int dim = trial_fe.GetDim(); // TODO: shouldn't it be the dimension of test_fe?
+	const int h1dofs_cnt = test_fe.GetDof();
+	const int l2dofs_cnt = trial_fe.GetDof();
+
+	if (equationV)
+	{
+		MFEM_VERIFY(rv.Size() == nqp, "");
+	}
+
+	if (equationE)
+	{
+		MFEM_VERIFY(re.Size() == nqp, "");
+		MFEM_VERIFY(v_e.Size() == h1dofs_cnt*dim, "");
+		MFEM_VERIFY(w_e.Size() == l2dofs_cnt, "");
+	}
+
+	DenseMatrix grad_vshape(h1dofs_cnt, dim); // grad of velocity basis vector
+	DenseMatrix loc_force(h1dofs_cnt, dim);
+	Vector Vloc_force(loc_force.Data(), h1dofs_cnt * dim);
+
+	for (int q = 0; q < nqp; q++)
+	{
+		const IntegrationPoint &ip = ir->IntPoint(q);
+
+		// Form stress:grad_vshape at the current point.
+		// TODO: how do we set which velocity basis vector is used for the
+		// gradient calculation?
+		test_fe.CalcDShape(ip, grad_vshape);
+		for (int i = 0; i < h1dofs_cnt; i++)
+		{
+			for (int vd = 0; vd < dim; vd++) // Velocity components.
+			{
+				loc_force(i, vd) = 0.0;
+				for (int gd = 0; gd < dim; gd++) // Gradient components.
+				{
+					loc_force(i, vd) +=
+						quad_data.stressJinvT(vd)(zone_id*nqp + q, gd) * grad_vshape(i,gd);
+				}
+			}
+		}
+
+		// NOTE: UpdateQuadratureData includes ip.weight as a factor in quad_data.stressJinvT,
+		// set by LagrangianHydroOperator::UpdateQuadratureData.
+		loc_force *= 1.0 / ip.weight;  // Divide by exact quadrature weight
+
+		if (equationV)
+		{
+			// set rv = sum(Vloc_force) 
+			rv[q] = 0.0;
+			for (int i = 0; i < h1dofs_cnt * dim; i++)
+			{
+				rv[q] += Vloc_force(i);
+			}
```

Comment:
For the velocity constraints I simply sum the `Vloc_force`, since the force vector for the velocity equation is just the integral of `stress : grad w_j`, with `w_j` a velocity basis vector. This is what's computed in the loop above.